lisp/ess-help.el

```diff
@@ -455,60 +455,47 @@ With (prefix) ALL non-nil, use `vignette(*, all=TRUE)`, i.e., from all installed
-  (catch 'win
-    (dolist (f (frame-list))
-      (when (frame-visible-p f)
-        (dolist (w (window-list f))
-          (when (with-current-buffer (window-buffer w) (derived-mode-p 'ess-help-mode))
-            (throw 'win w)))))))
-
-(defun ess--switch-to-help-buffer (buff)
-  "Switch to an ESS help BUFF.
-For internal use.  Take into account variable `ess-help-own-frame'."
-  (if (eq ess-help-own-frame t)
-      ;; 0) always pop new frame
-      (let* ((frame (make-frame ess-help-frame-alist))
-             (action `(display-buffer-same-window (reusable-frames . ,frame))))
-        (ess--display-help buff frame action))
-    (let ((help-win (or
-                     ;; if this doc is already displayed, reuse
-                     (get-buffer-window buff 'all-frames)
-                     ;; if help window visible, reuse
-                     (and ess-help-reuse-window
-                          (ess--find-displayed-help-window)))))
-      (cond
-       ;; 1) existent help window lying somewhere; reuse
-       (help-win
-        (set-window-buffer help-win buff)
-        (let ((action '(display-buffer-reuse-window (reusable-frames . t))))
-          (ess--display-help buff nil action)))
-       ;; 2)
-       ((eq ess-help-own-frame 'one)
-        (let* ((frame (if (frame-live-p ess--help-frame)
-                          ess--help-frame
-                        (make-frame ess-help-frame-alist)))
-               (action `(display-buffer-same-window (reusable-frames . ,frame))))
-          (setq ess--help-frame frame)
-          (ess--display-help buff frame action)))
-       ;; 3) display help in other window
-       (t
-        (ess--display-help buff))))))
-
-(defun ess--display-help (buff &optional frame action)
-  (let ((action (or action `(nil (reusable-frames . ,(or frame ess-display-buffer-reuse-frames))))))
-    (when (and (framep frame)
-               (not (eq (selected-frame) frame)))
-      ;; VS[04-05-2018]: `display-buffer` framework has no satisfactory
-      ;; functionality to display buffers in selected windows. So, do some extra
-      ;; frame selection here.
-      (raise-frame frame)
-      (if ess-help-pop-to-buffer
-          (select-frame-set-input-focus frame)
-        (select-frame frame)))
-    (if ess-help-pop-to-buffer
-        (pop-to-buffer buff action)
-      (display-buffer buff action))))
+(defun ess--display-help (buff)
+  "Display buffer BUFF.
+Call `ess-help-pop-to-buffer-function' to display the buffer. It
+gets called with two arguments, the buffer and an action alist.
+If `ess-help-pop-to-buffer-function' is not a function but is
+non-nil, use `pop-to-buffer', otherwise use `display-buffer'.
+Respects `display-buffer-alist' (and the other `display-buffer'
+framework variables), `ess-help-own-frame',
+`ess-help-frame-alist', and `ess-display-buffer-reuse-frames'.
+Rely either on the `display-buffer' variables or the ESS
+variables; do not try to use both to control help display."
```

Comment:
Mostly I'm referencing `display-buffer-alist` here, though there's also `display-buffer-base-action` and also a few obsolete variables. 

The last sentence is there because it's possible to customize things in such a way that you create conflicting settings (e.g. set `ess-display-buffer-reuse-frames` but then also include a `reusable-frames` entry in `display-buffer-alist`. I'm not sure what happens if you make conflicting settings and wanted to dissuade users from trying that out :-)